File: typescript/tests/giftcard/application/integration/payment/adapter/paymentResult.test.ts
import { describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import {
  paymentAcceptedIntegrationEvent,
  paymentExpiredIntegrationEvent,
  paymentRejectedIntegrationEvent,
} from '@/payment/integration/paymentResultIntegrationEvent.js';
import { PaymentResult } from '@/giftcard/application/integration/payment/adapter/paymentResult.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';

describe('PaymentResult', () => {
  const paymentResult = new PaymentResult();

  it('adapts PaymentAccepted to ConfirmTopUp', () => {
    const giftCardId = generateId((value) => new GiftCardId(value));
    const amount = new Money(50);
    const event = paymentAcceptedIntegrationEvent(giftCardId.value.value, amount);

    const result = paymentResult.adapt(event);

    expect(result).not.toBeNull();
    expect(result!.aggregateId).toEqual(giftCardId);
    expect(result!.amount).toEqual(amount);
  });

  it('adapts PaymentRejected to null', () => {
    const event = paymentRejectedIntegrationEvent(crypto.randomUUID(), new Money(50), 'declined');

    const result = paymentResult.adapt(event);

    expect(result).toBeNull();
  });

  it('adapts PaymentExpired to null', () => {
    const event = paymentExpiredIntegrationEvent(crypto.randomUUID(), new Money(50));

    const result = paymentResult.adapt(event);

    expect(result).toBeNull();
  });

  it('throws on null event', () => {
    expect(() => paymentResult.adapt(null as unknown as Parameters<typeof paymentResult.adapt>[0])).toThrow();
  });

  it('throws on null accepted event', () => {
    expect(() => paymentResult.adaptAccepted(null as unknown as Parameters<typeof paymentResult.adaptAccepted>[0])).toThrow();
  });
});

File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/integration/payment/adapter/PaymentResultTest.java
package it.giannibombelli.wsc2026.giftcard.application.integration.payment.adapter;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.application.commands.ConfirmTopUp;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.payment.integration.PaymentResultIntegrationEvent;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class PaymentResultTest {

    private final PaymentResult paymentResult = new PaymentResult();

    @Test
    void adaptPaymentAccepted_returnsConfirmTopUp() {
        GiftCardId giftCardId = EntityId.generate(GiftCardId::new);
        Money amount = new Money(new BigDecimal("50.00"));
        PaymentResultIntegrationEvent event = new PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent(
            giftCardId.value().toString(), amount
        );

        ConfirmTopUp result = paymentResult.adapt(event);

        assertThat(result).isNotNull();
        assertThat(result.aggregateId()).isEqualTo(giftCardId);
        assertThat(result.amount()).isEqualTo(amount);
    }

    @Test
    void adaptPaymentRejected_returnsNull() {
        PaymentResultIntegrationEvent event = new PaymentResultIntegrationEvent.PaymentRejectedIntegrationEvent(
            UUID.randomUUID().toString(), new Money(new BigDecimal("10.00")), "declined"
        );

        ConfirmTopUp result = paymentResult.adapt(event);

        assertThat(result).isNull();
    }

    @Test
    void adaptPaymentExpired_returnsNull() {
        PaymentResultIntegrationEvent event = new PaymentResultIntegrationEvent.PaymentExpiredIntegrationEvent(
            UUID.randomUUID().toString(), new Money(new BigDecimal("10.00"))
        );

        ConfirmTopUp result = paymentResult.adapt(event);

        assertThat(result).isNull();
    }

    @Test
    void adaptNullEvent_throwsException() {
        assertThatThrownBy(() -> paymentResult.adapt(null))
            .isInstanceOf(IllegalArgumentException.class);
    }
}
